### ISSUE TITLE
[java] UnnecessaryCast reported in stream chain `map()` call that casts to more generic interface

### ISSUE THREAD

**Issue by slovdahl** [pmd/pmd#5440]
**Affects PMD Version:** 7.9.0 (but not 6.55.0)

**Rule:** UnnecessaryCast

Please provide the rule name and a link to the rule documentation: https://pmd.github.io/pmd/pmd_rules_java_codestyle.html#unnecessarycast

**Description:**

Casting to a more generic interface in a stream pipeline's `map` method in order to satisfy the wanted generic type is reported as an uncessary cast by PMD.

**Code Sample demonstrating the issue:**

```java
package test;

import java.util.List;

class Test {
    interface Parent {}
    interface Child extends Parent {}

    void test() {
        List<Child> children = List.of();
        List<Parent> result = children.stream()
                .map(child -> (Parent)child)
                .toList();
    }
}
```

**Expected outcome:**

PMD reports a violation at line 12, but that's wrong. That's a false positive.

```
Test.java:12:     UnnecessaryCast:        Unnecessary cast (Parent)
```

**Running PMD through:** *Gradle 8.10*

**Comment by toKrause** [comment]
I stumbled upon, what I believe, is the same issue (casting to a more general type in order to ensure compatibility with the final result type in a fluently designed API):

```java
class Scratch {
    public static void main(String[] args) {
        Object object = Optional
            .of(new Object())
            .map(o -> (Object) o.toString())
            .orElse(Boolean.FALSE);
    }
}
```
